### PR TITLE
Support loras on quantized models

### PR DESCRIPTION
**Pull request by fmmoret**
https://github.com/vllm-project/vllm/issues/2821

Tried https://huggingface.co/yard1/llama-2-7b-sql-lora-test/blob/main/adapter_config.json against awq, gptq, & squeeze via editing https://github.com/vllm-project/vllm/blob/main/examples/multilora_inference.py

### CODE REVIEW

**Review by Yard1 — COMMENTED** [review]
Looking good! Could we add an integration test similar to what's in `tests/lora/test_llama.py`?

**Comment by Yard1** [comment]
@fmmoret I think as long as we can be sure that there are no exceptions and output is correct (even if it's not stable) it should be fine. You can limit it to stable prompts or reduce the number of tokens that are generated to see if it helps.

**Comment by samadkoita** [comment]
Is anyone planning to take this forward?

**Comment by samadkoita** [comment]
@fmmoret Is it okay if I take a crack at it & try some other examples here? Need this urgently for a project.

**Comment by Peter-Devine** [comment]
@fmmoret Any way I can help? This would be an awesome feature so I'd hate to see its inclusion falter.

**Comment by thincal** [comment]
could someone explain what's the idea behind for handling different dtype
of base and Lora weights?

**Comment by flexorRegev** [comment]
@fmmoret  @Yard1  From what it looks in this PR - there isn't anything inherent to the fact that quantized models + lora aren't supported right now - it just wasn't tested?

**Comment by hanan9m** [comment]
@fmmoret I can confirm that I was able to run a quantization model + adapters using your branch, and the results was good.

File: vllm/lora/layers.py [hunk 1]
```diff
@@ -668,25 +682,27 @@ def create_lora_weights(
...
-                self.base_layer.weight.shape[0],
+                self.base_layer.output_size,
                 lora_config.max_lora_rank,
             ),
             dtype=lora_config.lora_dtype,
-            device=self.base_layer.weight.device,
+            device=device,
         )
         self.indices: Optional[torch.Tensor] = None
         self.indices_len: Optional[List[int]] = None
```
Comment (jeejeelee) [inline vllm/lora/layers.py:708]: @fmmoret  Thank you for your contribution. I have tested this pull request, and the result looked good. However, one issue that needs to be considered is tensor parallelism.

**Comment by sunjunlishi** [comment]
> @fmmoret I can confirm that I was able to run a quantization model + adapters using your branch, and the results was good.

which  branch?

File: vllm/lora/layers.py (hunk 1 above)
Comment (desperadoola) [reply vllm/lora/layers.py:708]: Just change `self.base_layer.input_size` to `self.base_layer.input_size_per_partition` for `RowParallelLinear` module, and change `self.base_layer.output_size` to `self.base_layer.output_size_per_partition` for `ColumnParallelLinearWithLoRA`. We could run on tensor parallelism setting using this PR.

Comment (desperadoola) [reply vllm/lora/layers.py:708]: Also, there might raise NoneType error when tp_size > 1 and either lora_b[0] or lora_b[1] is None. It's still a bug in main branch

```python
if self.tp_size > 1:
            tensor_model_parallel_rank = get_tensor_model_parallel_rank()
            shard_size = self.output_dim
            start_idx = tensor_model_parallel_rank * shard_size
            end_idx = (tensor_model_parallel_rank + 1) * shard_size
            # lora_b = lora_b[0][:,
            #                    start_idx:end_idx], lora_b[1][:,
            #                                                  start_idx:end_idx]
            if lora_b[0] is not None:
                lora_b[0] = lora_b[0][:, start_idx:end_idx]
            if lora_b[1] is not None:
                lora_b[1] = lora_b[1][:, start_idx:end_idx]
```

**Pull request closed without merging**